SECEdgar/extractor/EDGARExtractor.py

```diff
@@ -0,0 +1,227 @@
+import logging
+import uu
+
+
+class EDGARExtractor():
+
+    # Compile regular expressions
+    # Note that re.search() does not have a starting position argument
+    # but the search() method of a Pattern object
+    # (a compiled regex) has a pos argument
+    re_doc = re.compile("<DOCUMENT>(.*?)</DOCUMENT>", flags=re.DOTALL)
+    re_sec_header = re.compile("<SEC-HEADER>.*?\n(.*?)</SEC-HEADER>",
+                               flags=re.DOTALL)
+    re_sec_doc = re.compile("<SEC-DOCUMENT>(.*?)</SEC-DOCUMENT>",
+                            flags=re.DOTALL)
+    re_text = re.compile("<TEXT>(.*?)</TEXT>", flags=re.DOTALL)
+
+    def __init__(self, in_file, out_dir=None, metadata_debug=False,
+                  create_subdir=True, rm_infile=False):
+        """
+        Extract the embedded documents and
+        their respective metadata from an EDGAR txt file
+        """
+        self.metadata_debug = metadata_debug
+        if out_dir is None:
+            out_dir = os.path.dirname(os.path.abspath(in_file))
+        self.procTxtFile(in_file, out_dir,
+                         create_subdir=create_subdir,
+                         rm_infile=rm_infile)
+
+    def procTxtFile(self, in_file, out_dir, create_subdir=True, rm_infile=False):
+
+        if os.path.splitext(in_file)[1] != ".txt":
+            raise Exception("ERROR: " + in_file + " is not a .txt file")
+
+        # Read input txt file
+        with open(in_file, encoding="utf8") as intxtfh:
+            intxt = intxtfh.read()
+
+        # Create a subdirectory with the same name of the input .txt file
+        if create_subdir:
+            out_dir = os.path.join(out_dir, os.path.basename(in_file))
+            os.makedirs(out_dir)
+
+        # Loop for every "<SEC-DOCUMENT>" in the file
+        sec_doc_cursor = 0
+        sec_doc_count = intxt.count("<SEC-DOCUMENT>")
+        for sec_doc_num in range(sec_doc_count):
+            # Extract the <SEC-DOCUMENT> part
+            sec_doc_m = self.re_sec_doc.search(intxt, pos=sec_doc_cursor)
+            if not sec_doc_m:
+                break
+            sec_doc_cursor = sec_doc_m.span()[1]
+            sec_document = sec_doc_m.group(1)
+
+            # Process metadata
+            metadata_txt_match = self.re_sec_header.search(sec_document)
+            metadata_txt = metadata_txt_match.group(1)
+            metadata_filename = "secdoc" + str(sec_doc_num) + ".metadata.json"
+            metadata_file = os.path.join(out_dir, metadata_filename)
+            metadata = self.processMetadata(metadata_txt)
+            logging.info("Metadata written into {}".format(metadata_file))
+
+            # Loop through every document
+            metadata["documents"] = list()
+            documents = sec_document[metadata_txt_match.span()[1]:].strip()
+            doc_count = documents.count("<DOCUMENT>")
+            doc_cursor = 0
+            for doc_num in range(doc_count):
+                # Pattern.search(string[, pos[, endpos]])
```

Comment:
You can also break this into methods. It is hard to understand what is exactly going on here.

each of these operations can be a collection of small functions inside a util class.

like
```
class ExtractorHelper():
            def header_from(cost_section)
            def table_from(header)
```

Reply:
ok

Reply:
I agree with that